app/scripts/minmax/dimMinMax.controller.js

```diff
@@ -169,27 +320,127 @@
       ranked: {},
       filter: {
         int: 3,
         dis: 2,
         str: 2
       },
+      lockedItemsValid: function(dropped_id, dropped_type) {
+          dropped_id = dropped_id.split('-')[1];
+          var item = _.findWhere(buckets[vm.active][dropped_type], {id: dropped_id});
+          var exoticCount = (item.tier === 'Exotic')? 1 : 0;
+          for(var type in vm.lockeditems) {
+              var item = vm.lockeditems[type];
+              if(item === null || type === dropped_type) { continue; }
+              if(item.tier === 'Exotic' && item.type != 'ClassItem') {
+                  exoticCount += 1;
+              }
+          }
+          return exoticCount < 2;
+      },
+      onOrderChange: function () {
+        vm.setOrderValues = vm.setOrder.split(',')
+      },
+      onDrop: function(dropped_id, type) {
+          dropped_id = dropped_id.split('-')[1];
+          var item = _.findWhere(buckets[vm.active][type], {id: dropped_id});
+          vm.lockeditems[type] = item;
+          var bestarmor = getBestArmor(buckets[vm.active], vm.lockeditems); 
+          vm.highestsets = vm.getSetBucketsStep(bestarmor);
+      },
+      onRemove: function(removed_type) {
+          vm.lockeditems[removed_type] = null;
+          var bestarmor = getBestArmor(buckets[vm.active], vm.lockeditems); 
+          vm.highestsets = vm.getSetBucketsStep(bestarmor);
+      },
+      getSetBucketsStep: function(bestArmor) {
+            var helms = bestArmor['helmet'];
+            var gaunts = bestArmor['gauntlets'];
+            var chests = bestArmor['chest'];
+            var legs = bestArmor['leg'];
+            var classItems = bestArmor['classItem'];
+            var ghosts = bestArmor['ghost'];
+            var artifacts = bestArmor['artifact'];
+            
+            var set_map = {};
+            var get_val = function(stats, type, target_type) {
+                return stats.scaled + ((type == target_type)? stats.bonus : 0);
+            };
+            
+            function step(h, g, c, l, ci, gh, ar, processed_count) {
+                for(; h < helms.length; ++h) {
+                for(; g < gaunts.length; ++g) {
+                for(; c < chests.length; ++c) {
+                for(; l < legs.length; ++l) {
+                for(; ci < classItems.length; ++ci) {
+                for(; gh < ghosts.length; ++gh) {
+                for(; ar < artifacts.length; ++ar) {
+                    var armor = {helmet: helms[h], gauntlets: gaunts[g], chest: chests[c], leg: legs[l], classItem: classItems[ci], ghost: ghosts[gh], artifact: artifacts[ar]};
+                    if(validSet(armor)) {
+                        var set = {armor: armor};
+                        set.int_val =  get_val(set.armor.helmet.item.normalStats[0], set.armor.helmet.bonus_type, 'int') + get_val(set.armor.gauntlets.item.normalStats[0], set.armor.gauntlets.bonus_type, 'int') + get_val(set.armor.chest.item.normalStats[0], set.armor.chest.bonus_type, 'int') +
```

Comment:
sometimes the API will return only two stats on an item (only int/dis), so normalItems[2] might actually return null. We might also not want to rely on the index in that case.

Instead maybe something similar to what you did above:

quick example (this is working for me, though it's probably possible to cut down on the LoC and use a function instead):
```
                        set.int_val =
                          get_val(_.findWhere(set.armor.helmet.item.normalStats, {statHash: 144602215}) || {scaled: 0, bonus: 0}, set.armor.helmet.bonus_type, 'int') +
                          get_val(_.findWhere(set.armor.gauntlets.item, {statHash: 144602215}) || {scaled: 0, bonus: 0}, set.armor.gauntlets.bonus_type, 'int') +
                          get_val(_.findWhere(set.armor.chest.item, {statHash: 144602215}) || {scaled: 0, bonus: 0}, set.armor.chest.bonus_type, 'int') +
                          get_val(_.findWhere(set.armor.leg.item, {statHash: 144602215}) || {scaled: 0, bonus: 0}, set.armor.leg.bonus_type, 'int') +
                          get_val(_.findWhere(set.armor.classItem.item, {statHash: 144602215}) || {scaled: 0, bonus: 0}, set.armor.classItem.bonus_type, 'int') +
                          get_val(_.findWhere(set.armor.ghost.item, {statHash: 144602215}) || {scaled: 0, bonus: 0}, set.armor.ghost.bonus_type, 'int') +
                          get_val(_.findWhere(set.armor.artifact.item, {statHash: 144602215}) || {scaled: 0, bonus: 0}, set.armor.artifact.bonus_type, 'int');

                        set.disc_val =
                          get_val(_.findWhere(set.armor.helmet.item.normalStats, {statHash: 1735777505}) || {scaled: 0, bonus: 0}, set.armor.helmet.bonus_type, 'disc') +
                          get_val(_.findWhere(set.armor.gauntlets.item, {statHash: 1735777505}) || {scaled: 0, bonus: 0}, set.armor.gauntlets.bonus_type, 'disc') +
                          get_val(_.findWhere(set.armor.chest.item, {statHash: 1735777505}) || {scaled: 0, bonus: 0}, set.armor.chest.bonus_type, 'disc') +
                          get_val(_.findWhere(set.armor.leg.item, {statHash: 1735777505}) || {scaled: 0, bonus: 0}, set.armor.leg.bonus_type, 'disc') +
                          get_val(_.findWhere(set.armor.classItem.item, {statHash: 1735777505}) || {scaled: 0, bonus: 0}, set.armor.classItem.bonus_type, 'disc') +
                          get_val(_.findWhere(set.armor.ghost.item, {statHash: 1735777505}) || {scaled: 0, bonus: 0}, set.armor.ghost.bonus_type, 'disc') +
                          get_val(_.findWhere(set.armor.artifact.item, {statHash: 1735777505}) || {scaled: 0, bonus: 0}, set.armor.artifact.bonus_type, 'disc');

                        set.str_val =
                          get_val(_.findWhere(set.armor.helmet.item.normalStats, {statHash: 4244567218}) || {scaled: 0, bonus: 0}, set.armor.helmet.bonus_type, 'str') +
                          get_val(_.findWhere(set.armor.gauntlets.item, {statHash: 4244567218}) || {scaled: 0, bonus: 0}, set.armor.gauntlets.bonus_type, 'str') +
                          get_val(_.findWhere(set.armor.chest.item, {statHash: 4244567218}) || {scaled: 0, bonus: 0}, set.armor.chest.bonus_type, 'str') +
                          get_val(_.findWhere(set.armor.leg.item, {statHash: 4244567218}) || {scaled: 0, bonus: 0}, set.armor.leg.bonus_type, 'str') +
                          get_val(_.findWhere(set.armor.classItem.item, {statHash: 4244567218}) || {scaled: 0, bonus: 0}, set.armor.classItem.bonus_type, 'str') +
                          get_val(_.findWhere(set.armor.ghost.item, {statHash: 4244567218}) || {scaled: 0, bonus: 0}, set.armor.ghost.bonus_type, 'str') +
                          get_val(_.findWhere(set.armor.artifact.item, {statHash: 4244567218}) || {scaled: 0, bonus: 0}, set.armor.artifact.bonus_type, 'str');
```